File: 3Sum.java
/*
Runtime: 21 ms, faster than 99.57% of Java online submissions for 3Sum.
Memory Usage: 46.6 MB, less than 92.96% of Java online submissions for 3Sum.
*/

//Optimized Solution TC: O(n^2)
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        //Sorting The Array For Applying 2 Pointer Method
        Arrays.sort(nums);
        
        // Creating the Result List To Be Returned
        List<List<Integer>> result = new LinkedList<>();
        
        // Outer Loop To Traverse Through All Elements And Stopping It Before The Last 2 Elements
        for(int i=0; i<nums.length-2; i++){
            
            // Condition To Avoid Duplicates In The Result Array
            if(i==0 || (i>0 && nums[i] != nums[i-1])){
                
                // Taking The Start As The Next Element To The Index Of Element In The Outer Loop
                // Taking The End As The Index Of The Last Element In The Array
                // This Works Only Because The Array Is Sorted.
                int start = i+1;
                int end = nums.length-1;
                
                // To Get The Answer As Zero, We Set The Sum As -nums[i];
                int sum = 0 - nums[i];
                
                // 2 Pointers Traversing From the Beginning And End Of The Subarray
                while(start < end){
                    
                    /*  nums[start] + nums[end] = 0 - nums[i] 
                        equates to 
                        nums[start] + nums[end] + nums[i] == 0 */
                    if(nums[start] + nums[end] == sum){
                        
                        // Adding Result To The Result Array
                        result.add(Arrays.asList(nums[i], nums[start], nums[end]));
                        
                        // Skipping Duplicate Elements In A Sorted Array From The Start
                        while(start < end && nums[start] == nums[start+1]){
                            start++;
                        }
                        // Skipping Duplicate Elements In A Sorted Array From The End
                        while(start < end && nums[end] == nums[end-1]){
                            end--;
                        }
                        // Skipping The Last Duplicates From Both Sides
                        start++;
                        end--;
                        
                    }
                    // Moving Pointers From Both Sides Towards Mid Of The Array
                    else if(nums[start] + nums[end] < sum){
                        start++;
                    }else{
                        end--;
                    }
                }
            }
        }
        
        // Returning The Result Array
        return result;
        
    }
}

File: 3Sum.js
/**
 * This function finds all unique triplets in the array which gives the sum of zero.
 * @param {number[]} nums - An array of integers.
 * @return {number[][]} - A list of triplets, where each triplet is an array of three integers that add up to zero.
 */
// TC: O(N^2) SC: O(1)
var threeSum = function(nums) {
    // Sort the input array in ascending order to make it easier to find triplets.
    nums.sort((a,b) => a - b);
    const result = [];
    // Iterate through the array, considering each number as a potential first number of the triplet.
    for(let i = 0; i < nums.length - 2; i++) {
        // If the current number is greater than zero, break from the loop as further numbers will also be greater than zero and cannot sum to zero.
        if(nums[i] > 0) break;
        // Avoid duplicates: skip the same number to prevent duplicate triplets.
        if(i === 0 || (i > 0 && nums[i] !== nums[i - 1])){
            // The target sum for the other two numbers is the negative of the current number.
            const target = 0 - nums[i];
            // Initialize two pointers, one at the next element and the other at the end of the array.
            let start = i + 1, end = nums.length - 1;
            // Move the pointers towards each other until they meet.
            while(start < end) {
                // Calculate the sum of the numbers at the two pointers.
                const sum = nums[start] + nums[end];
                // If the sum equals the target, we found a valid triplet.
                if(sum === target) {
                    // Add the triplet to the result array.
                    result.push([nums[i], nums[start], nums[end]]);
                    // Skip over duplicates for the second number.
                    while(start < end && nums[start] === nums[start + 1]) start++;
                    // Move the start pointer to the next unique number.
                    start++;
                    // Skip over duplicates for the third number.
                    while(start < end && nums[end] === nums[end - 1]) end--;
                    // Move the end pointer to the previous unique number.
                    end--;
                }
                // If the sum is less than the target, move the start pointer to the right to increase the sum.
                else if(sum < target) {
                    while(start < end && nums[start] === nums[start + 1]) start++;
                    start++;
                }
                // If the sum is greater than the target, move the end pointer to the left to decrease the sum.
                else {
                    while(start < end && nums[end] === nums[end - 1]) end--;
                    end--;
                }
            } 
        }
    }
    // Return the array of triplets that sum up to zero.
    return result;
};
